connectors/spark/src/main/scala/io/unitycatalog/spark/auth/GenericCredentialProvider.java

```diff
@@ -123,31 +132,82 @@ private GenericCredential renewCredential() throws ApiException {
     }
   }
 
+  private boolean isRecoverable(Throwable e) {
+    if (e instanceof ApiException) {
+      int code = ((ApiException) e).getCode();
+      // Retry on rate limit (429), service unavailable (503), and 5xx server errors
+      return code == 429 || code == 503 || (code >= 500 && code < 600);
+    }
+    // Network-level transient failures
+    return e instanceof java.net.SocketTimeoutException  // Timeout
+      || e instanceof java.net.SocketException           // Connection issues
+      || e instanceof java.net.UnknownHostException;     // DNS resolution issues
+  }
+
+  private TemporaryCredentials callWithRetry(ApiCallSupplier apiCall) throws ApiException {
+    int maxAttempts = conf.getInt(UCHadoopConf.RETRY_MAX_ATTEMPTS_KEY, UCHadoopConf.RETRY_MAX_ATTEMPTS_DEFAULT);
+    long initialDelay = conf.getLong(UCHadoopConf.RETRY_INITIAL_DELAY_KEY, UCHadoopConf.RETRY_INITIAL_DELAY_DEFAULT);
+    double multiplier = conf.getDouble(UCHadoopConf.RETRY_MULTIPLIER_KEY, UCHadoopConf.RETRY_MULTIPLIER_DEFAULT);
+
+    Exception lastException = null;
+
+    for (int attempt = 1; attempt <= maxAttempts; attempt++) {
+      try {
+        return apiCall.get();
+      } catch (Exception e) {
+        lastException = e;
+
+        if (!isRecoverable(e) || attempt == maxAttempts) {
+          break;
+        }
+
+        long baseDelay = (long) (initialDelay * Math.pow(multiplier, attempt - 1));
+        double jitter = (Math.random() - 0.5) * 2 * UCHadoopConf.RETRY_JITTER_FACTOR;
+        long delay = (long) (baseDelay * (1 + jitter));
+
+        try {
+          Thread.sleep(delay);
```

Comment:
Maybe it will be better to use the abstracted clock, so that we can easily do the clock simulation for testing purposes. 

https://github.com/unitycatalog/unitycatalog/pull/1114/files